### PR TITLE
Upstream Configuration

### PR DESCRIPTION
**Pull request by beverlylytle**
`Configuration` believe the grass is greener over here in `compas`.  

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/compas/robots/configuration.py [hunk 1]
```diff
@@ -0,0 +1,614 @@
...
+        joint_types = [_type_dict[name] for name in joint_names]
+
+        return Configuration(joint_values, joint_types, joint_names)
+
+
+if __name__ == "__main__":
+    import math  # noqa F401
+    from compas.geometry import allclose   # noqa F401
+    import doctest
+    doctest.testmod(globs=globals())
```
Comment (gonzalocasas) [inline src/compas/robots/configuration.py]: I think we can get rid of these blocks

File: src/compas/robots/__init__.py [hunk 2]
```diff
@@ -115,6 +116,10 @@
 """
 from __future__ import absolute_import
 
+from .configuration import (
+    Configuration,
+    FixedLengthList
```
Comment (gonzalocasas) [inline src/compas/robots/__init__.py]: Should we expose this class or rather keep it an implementation detail?

File: src/compas/robots/model/robot.py [hunk 3]
```diff
@@ -704,7 +766,7 @@ def compute_transformations(self, joint_state, link=None, parent_transformation=
 
         Parameters
         ----------
-        joint_state : dict
+        joint_state : dict or :class:`compas.robots.Configuration`
```
Comment (gonzalocasas) [inline src/compas/robots/model/robot.py]: totally minor detail, but I'd swap the order of these to make `Configuration` the perceived preferred option

**Review by gonzalocasas — APPROVED** [review]
LGTM! We've been through the details of the impl on compas_fab repo, so I only checked the stuff outside the config class.

File: src/compas/robots/__init__.py (hunk 2 above)
Comment (beverlylytle) [reply src/compas/robots/__init__.py]: Yeah, this is tricky... I use it in `JointTrajectoryPoint`.

Comment (beverlylytle) [reply src/compas/robots/__init__.py]: Should I make it available as a third level import rather than second?

Comment (gonzalocasas) [reply src/compas/robots/__init__.py]: yes, that sounds like a good alternative

**Pull request merged**